### DISCUSSION TITLE
Help with model selection, protein mixture models

### DISCUSSION THREAD

**Discussion by veronicamb** [iqtree/iqtree2#200] [General]
Hello all, 

I'm trying to use iqtree to find the best-fitting model for a protein dataset and I'd like to include Protein Mixture Models in the models considered.  

Here's the command I used:

iqtree -s ../04_alignments/03_trimmed/06_informed/01_coding_NUC/01_AA \
       -pre Subset1_test2 \
       -nt AUTO \
       -ntmax 40 \
       -m MF \
       -mset WAG,LG,JTT \
       -msub nuclear \
       -madd C10,C20,C30,C40,C50,C60,LG4M,LG4X,CF4

However, the protein mixture models were not considered during the search. Can you help with what's missing in the command to fix it?

thank you!

Veronica

**Comment by roblanf** [comment]
@veronicamb here's one way to do what you want. I think the issue is that you have some competing demands in `mset` and `madd`, and `-mset` overrides `-madd`. 

So, to fix that you can just put all the models you are interested in in `-mset`, like this:

```
iqtree2 -s test.fas -m MF -mset WAG,LG,JTT,C10,C20,C30,C40,C50,C60,LG4M,LG4X,CF4
```

* `-m MF` suppresses tree search
* `-mset ...` specifies the models to use in MF

Another tip - if you know you have 40 threads available, then I suggest just specifying all 40 for ModelFinder. It will usually be faster than `-nt AUTO`.

Just to note, I have to say that I too am wondering if we have a bug with `-madd`, because I can't make it behave as expected. We are working on this as we speak. The number of models has exploded in the last few years, so it's time for a re-organisation of how they are presented and analysed. In fact, we have working on this all afternoon today. Maybe @bqminh can put me straight on that though!

**Reply by veronicamb** [reply]
Thank you so much! This is a huge help. I'll try that right away.

Do you think there's a difference in the behavior of -madd in different versions of iqtree? I'm using 2.2.2

**Reply by roblanf** [reply]
There shouldn't be a difference. I tested with 2.3.4, but I have to say it didn't work as I thought it should. Best thing is to wait 24 hrs - we are meeting about this literally today. We are finalising MF2, and that requires some rationalisation of how models are presented.

**Reply by bqminh** [reply]
There is a bug that -madd is ignored (whereas IQ-TREE 1 works as expected). I don't know when this bug was "introduced", but @thomaskf will fix it. Thanks

**Reply by thomaskf** [reply]
@veronicamb I tested one protein dataset using the latest version, v2.3.4, with the exact command you used (except the number of CPU threads), and I found the option "-madd" should work. During the computation of the best model, the mixture models C20,C30,...,C60,LG4M,LG4X, and CF4 were considered. I am attaching a log file for your reference. As you can see, those mixture models were considered. Please feel free to let me know if you have any questions.
[protein.phy.log](https://github.com/iqtree/iqtree2/files/15427381/protein.phy.log)

**Reply by veronicamb** [reply]
Thank you! I've been using iqtree 1, but I'll definitely give this a shot!

**Comment by veronicamb** [comment]
Will do! Thanks again

**Comment by 000generic** [comment]
I have a similar question - and it looks like its not solved in protein.phy.log file posted above.

I'd like to use ModelFinder (-m MF) to identify the best fit model for thousands of gene alignments to then use for a species tree - providing a partition file including the ModelFinder best fit models per partition of the concatenated alignment for IQTree generation of the species tree - and in parallel generate gene trees by IQTree after model selection ( so actually run -MFP) for use in Asteroid generation of the species tree.  

In doing this, I would like ModelFinder to consider the base models in IQTree (or even just the nuclear base models) - and 8 additional models I generated from a subset of the gene alignments using IQTree.

I would also like ModelFinder to consider mixture models C10, C20, and C60 - so for example, testing both the base model LG and LG combined with C20, LG+C20.

My command line is:

`iqtree2 -m MF -madd C10,C20,C60 -s mollusca___4767at6447___subtree1.aa --prefix output/mollusca___4767at6447___subtree1 -mset Blosum62,Dayhoff,DCMut,EAL,ELM,JTT,JTTDCMut,LG,NQ.bird,NQ.insect,NQ.mammal,NQ.pfam,NQ.plant,NQ.yeast,Poisson,PMB,Q.bird,Q.insect,Q.mammal,Q.pfam,Q.plant,Q.yeast,VT,WAG,input/nQ.mollusca42_c50_s50_m1,input/nQ.mollusca42_c50_s50_m2,input/nQ.mollusca42_c50_s75_m1,input/nQ.mollusca42_c50_s75_m2,input/Q.mollusca42_c50_s50_m1,input/Q.mollusca42_c50_s50_m2,input/Q.mollusca42_c50_s75_m1,input/Q.mollusca42_c50_s75_m2 -mwopt -T 5  --threads-max 5 --symtest-remove-bad
`

I'm unclear from the discussion above if -mset still overrides -madd but it seems to do so in various output where I have been testing parameters in IQTree (version 2.3.6). 

Currently, my solution in testing things is to add all combinations into -mset - so something like:

`-mset LG,LG+C20 
`

and not 

`-madd C20 -mset LG OR -mset LG,C20
`

I'm wondering if there is a more elegant way to do this - as its going to get ugly with so many base and mixture model combinations to consider. 

More generally, I am wondering if I might be do something wrong in the design of my approach or command line.

Thank you very much :) Eric

**Reply by evolbeginner** [reply]
Hi Eric,

I see this is a very good suggestion and i think your solution is correct. it's this post that brings me here https://bsky.app/profile/andrewjroger.bsky.social/post/3mndaw7fvec2z

perhaps this could bring the attention of the deveopers to this issue (or perhaps it's designed to run this way)!

sishuo

**Comment by roblanf** [comment]
Thanks for all this feedback. We're in the process over the next couple of months of changing and clarifying how models are specified in IQ-TREE, precisely to clear this kind of confusion up. The issue is that the set of available models has changed quite dramatically since IQ-TREE was initially designed. 

As an example motivated by the confusion above - it's not always obvious what a user wants with the current setup. This is because models have 3 very different components: the rates (like +G, +R, etc), the frequencies, and the exchangeabilities. In some cases, the frequencies and exchangeabilities go together (e.g. the LG model comes with its own frequencies, CXX models were estimated with POISSON exchangeabilities), some don't (MEOW profile mixtures for example). Some are more complex than that (e.g. LG4X and LG4M). Then there are things that seem sensible but are inadvisable, like C60+F models. 

In case anyone on this thread wants to provide feedback, the plan is to separate these into three things: rates, exchangeabilities, and frequencies. Then users can determine in a fine-grained way exactly what they want. E.g. three parameters would define the full model set:

* *mexchange* is the list of R matrices (e.g. LG, Q.PFAM, GTR, etc). Can include sensible lists like 'nuclear', 'viral' 'mrbayes', or whatever. 
* *mrate* RHAS models like +I, +I+G, +R, etc
* *mfreq* F vectors (e.g. +F, LG, Q.PFAM, C60, UDM8 etc)

The intention is to then fully cross these three groups to arrive at the final model list, but crucially excluding models that should not be used, such as LG4X+F, C60+F etc. We can include an override for people that insist on using models that the experts don't like, because we appreciate that sometimes this is important for research! 

We will additionally add some fine-grained control for the different parts of model selection, since it's typically not necessary (or sensible, if you care about not using unnecessary compute) to compare ALL models at ALL stages. The three stages are:

* Initial model to get the guide tree (we'll allow users to specify this, and have sensible defaults to general models)
* Models used for merging partitions
* The full model set to test on the final partitioning scheme

Anyone who'd like to provide feedback on this is very much encouraged to do so!

Rob

**Reply by evolbeginner** [reply]
Thx so much Rob! What you've been working on seems to me to be fantastic. I also think mfreq,mrate,mexchange would greatly clarify the use of diff subs models, which are often used in a somehow confusing way in papers and similarly, this is also what eq. 1 in [this](https://academic.oup.com/mbe/article/43/3/msag039/8489165) (sorry for shameless self-promotion) (see [Guindon 2019](https://gensoft.pasteur.fr/docs/phyml/3.3.20190909/phyml-manual.pdf) for the original formula, pp 34-35) hopes to make clear.

I guess an important thing that users would keep in mind now is the inclusion of complex subs models at least for a test. If so, currently, it seems that sth like `-mset LG,LG+C60 -mrate E,G,I,G+I` makes a good balance for current settings otherwise if Cxx is auto included whenever `-m TESTONLY` is specified, it could cost much time on large data sets. You might also wish to see [this](https://sishuowang2022.wordpress.com/2023/06/20/substitution-models-lgcxx-and-cxxlg-differ-in-iq-tree/) for Cxx+LG vs. LG+Cxx although it seems a rather trivial issue now.

Thank you!

cheers,
sishuo
